### PR TITLE
Test StringHierarchy Area and Location Factories

### PR DESCRIPTION
**Pull request by cld99**

### CODE REVIEW

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public StringHierarchyAreaFactory(StringHierarchyRegion baseModelRegion)
 
         this.HelpMsg = $"A StringHierarchyArea can be built from one of the " +
             $"following path specs, selecting one option from any list in [] " +
-            $"and excluding the []:\n{baseModelRegion.ToString()}";
+            $"and excluding the []:\n{baseModelRegion?.ToString() ?? "null"}";
```
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs]: In which cases were you finding baseModelRegion was null? It should always have a value after object initialization, so we don't really want to allow the possibility that it's null.

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs [hunk 2]
```diff
@@ -34,12 +34,12 @@ private bool AreaSpecCheckFunction(string fullSpec)
         var parts = fullSpec.Split(
             StringHierarchySpec.REGIONS_MODIFIERS_DELIMITER);
 
-        return parts.Length > 0 &&
+        return parts.Length < 3 &&
```
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs]: For which cases was this change needed? What was failing? May be a larger bug/flaw

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs [hunk 3]
```diff
@@ -34,12 +34,12 @@ private bool AreaSpecCheckFunction(string fullSpec)
...
             // First try to navigate this model to the specified region.
             this._baseRegion.TryGetSubregion(parts[0], out var subregion) &&
             // Then - if any given - check if the modifiers valid in the model.
-            (parts.Length > 1) ?
-            subregion.IsValidModifierSpec(parts[1]) : true;
+            ((parts.Length > 1) ?
+            subregion.IsValidModifierSpec(parts[1]) : true);
```
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs]: Just curious: did adding the parens change the functionality and fix something or was this just a style clean?

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:1]: Same comments/questions in this file as in AreaFactory

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs:1]: Please add doc comments to the source class, properties, and methods too

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        this._output = testOutputHelper;
+    }
+
+    // Test method naming convention: LibClassMethodName_ScenarioShouldExpectn
+
+    /// <summary>
+    /// Test the constructor with null parameter.
+    /// </summary>
+    [Fact]
+    public void Constructor_ShouldAcceptNull()
```
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs]: We actually want to disallow null input. Is the current implementation (pre your related changes in the source code files above) allowing it?

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        Assert.NotNull(product);
+        Assert.Equal((IArea) new StringHierarchyArea(""), product);
+    }
+
+    /// <summary>
+    /// Test the TryCreate method with paramValues including a fullSpec value
+    /// and not skipping value validation
+    /// </summary>
+    [Fact]
+    public void TryCreate_ShouldValidateShouldMaybeCreate()
```
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs]: What is the "ShouldMaybe"? Are there some test values for which it should vs would not create? In that case, put them in separate test functions. This will be made easier by using [Theory] rather than [Fact] flags and then just having a bunch of different [InlineData()] test input options or all of them in one [MemberData()] flag. See some of the test classes merge into the hardware or release branch for examples. Restructuring to use these may or may not change which test methods you actually have, e.g., giving you one test method per condition tested, e.g., TryCreate_ValidatedValidParamsShouldCreate, TryCreate_UnvalidatedValidParamsShouldCreate, TryCreate_ValidatedInvalidParamsShouldFail. Then the different input values you give are different examples of each condition

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs:1]: Same comments as in AreaFactory tests

File: docs/images/association_arrows-techvidvan.png
Comment (lhmcgann) [inline docs/images/association_arrows-techvidvan.png:1]: Are you by chance able to undo these changes to the image files? They have already been pushed in release which I know you aren't merging into yet, so if not possible it's alright we'll deal with it later, but if you can it will help prevent merge conflicts later.

**Review by lhmcgann — CHANGES_REQUESTED** [review]

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs (hunk 1 above)
Comment (johnshi314) [reply src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs]: I think baseModelRegion will only be null if it's passed to the constructor as null. If it should never be null, it would be better to throw an exception instead.

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs (hunk 4 above)
Comment (johnshi314) [reply tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs]: The modified constructor treats null as an acceptable parameter. If we don't want this we should throw an exception in the constructor instead.

**Review by johnshi314 — COMMENTED** [review]
If null shouldn't be an acceptable parameter then an exception should be thrown if we try to pass null to it. Also it would be more clear to seperate the test cases in the TryCreate_ShouldValidateShouldMaybeCreate to seperate when it should create from when it shouldn't to reduce ambiguity.

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs (hunk 3 above)
Comment (cld99) [reply src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs]: Yes, it first caused an error with the empty string test at lines 249-251 of StringHierarchyAreaFactoryTests. It looks like the conditional operator has lower precedence than &&, so assuming that we don't want all of the previous checks to default the output to true if they are false, the parentheses are necessary.

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs (hunk 2 above)
Comment (lhmcgann) [reply src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactory.cs]: per discussion: include both this and > 0

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs [hunk 6]
```diff
@@ -28,36 +51,67 @@ public StringHierarchyLocationFactory(
...
+    /// </summary>
+    /// <param name="fullSpec">The full location spec to validate.</param>
+    /// <returns>True if valid location spec, false if not.</returns>
+    private bool _LocationSpecCheckFunction(string fullSpec)
     {
         var parts = fullSpec.Split(
             StringHierarchySpec.REGIONS_MODIFIERS_DELIMITER);
 
-        return parts.Length > 0 &&
+        return parts.Length > 0 && parts.Length < 3 &&
```
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:73]: Use StringHierarchySpec.ParseFullSpec() here (and in the corresponding Area function) since that is the fullSpec parse and validation logic. I believe that logic just ignores any extra parts rather than erroring, so want to stay consistent.

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs [hunk 7]
```diff
@@ -28,36 +51,67 @@ public StringHierarchyLocationFactory(
...
+    /// <param name="product">An output parameter: the StringHierarchyLocation
+    /// if created, else null.</param>
+    /// <param name="skipValueValidation">An optional parameter: perform only
+    /// basic input validation and skip any deeper parameter value validation,
+    /// e.g., if the data has already been validated. Default: false.</param>
+    /// <returns>True if a StringHierarchyLocation could be created, False if 
+    /// not.</returns>
     public bool TryCreate(Dictionary<string, object> paramValues,
-        out ILocation product, bool skipValueValidation = false)
+        out ILocation? product, bool skipValueValidation = false)
```
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:98]: Added ? to indicate this output param could be null, an added ! in the function body to suppress null warnings (since the function value is passed to can handle null values, or value will have been checked to be non-null by the time it has used). You will need to similarly use ? and ! in the test classes to declare variables that could be null (i.e., your variables for output params) and suppress null value warnings (in my VSCode, all the places warnings appear show as yellow-underlined which could help identify them). I started changing some places in the SHLocationFactoryTests.cs class, but only barely --> will need to do the rest

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    /// </summary>
+    [Fact]
+    public void Constructor_Null_ShouldError()
+    {
+        bool caughtError = false;
+        try
+        {
+            // Create factory with null base region
+            StringHierarchyAreaFactory factory = 
+                new StringHierarchyAreaFactory(null!);
```
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyAreaFactoryTests.cs]: Example of ! use

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        // Create factory with non-null base region
+        StringHierarchyLocationFactory factory =
+            new StringHierarchyLocationFactory(region);
+
+        // Create dictionary of parameter values
+        Dictionary<string, object> paramValues =
+            new Dictionary<string, object>();
+
+        // Create variable to store produced IArea
+        ILocation? product;
```
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs]: Example of ? use

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+        // Check that error occurred
+        Assert.True(caughtError);
+    }
+
+    /// <summary>
+    /// Test the constructor with non-null parameter.
+    /// </summary>
+    [Fact]
+    public void Constructor_NonNull_ShouldInit()
```
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs:49]: If it doesn't make the test method name too long, try keeping to this <MethodName>_<Scenario>_Should<ExpectedBehavior/Output> format

File: tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        // Create region with the above parameters
+        var region = new StringHierarchyRegion("base", null,
+            null, regionOptions, null, regionSubregions);
+
+        // Create factory with non-null base region
+        StringHierarchyLocationFactory factory =
+            new StringHierarchyLocationFactory(region);
+
+        // Create dictionary with valid spec
+        Dictionary<string, object> paramValues1 =
```
Comment (lhmcgann) [inline tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactoryTests.cs]: Since you've written all the test cases this way already, fine for this PR, but in future, if you find yourself testing a whole bunch of different variations of data within the same test method, try to use a [Theory] tag with [InlineData()] or [MemberData()] instead

**Review by lhmcgann — CHANGES_REQUESTED** [review]

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs (hunk 6 above)
Comment (cld99) [reply src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:73]: ParseFullSpec parses the component region and modifier specs, however this is already handled within the calls to TryGetSubregion and IsValidModifierSpec respectively, so unless we want to reconstruct the region and modifier specs, this will not be feasible. This function specifically errors if extra parts are included because (at least as specified in the doc comments) the spec passed in must contain a single region and at most one modifier set.

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs [hunk 12]
```diff
@@ -7,16 +7,39 @@
...
+    /// "fullSpec" which must be a valid full string location spec wihtin the
+    /// base region.
+    /// </summary>
     public Dictionary<string, IDataLimits> ParamLimits { get; init; }
 
+    /// <summary>
+    /// Create an instance of the factory using the supplied base region to 
+    /// initialize help and data validation properties.
+    /// </summary>
+    /// <exception cref="NullReferenceException">Thrown if the base region is
```
Comment (lhmcgann) [inline src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:41]: what within this throws the exception? Is it just the ToString() call in generating the help msg?

**Review by lhmcgann — APPROVED** [review]
Great job! Just made a few organizational edits (used MemberData tags to condense and make a little more readable) and have one curiosity question, but good to go!

File: src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs (hunk 12 above)
Comment (cld99) [reply src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:41]: Yes, it is just the ToString() call if a null baseModelRegion is passed in.

Comment (lhmcgann) [reply src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyLocationFactory.cs:41]: yes :) could add an explicit null check in future
